daemon/snap.go

```diff
@@ -168,27 +170,140 @@ type screenshotJSON struct {
 	Height int64  `json:"height,omitempty"`
 }
 
-func mapLocal(about aboutSnap) map[string]interface{} {
-	localSnap, snapst := about.info, about.snapst
-	status := "installed"
-	if snapst.Active && localSnap.Revision == snapst.Current {
-		status = "active"
+type bySnapApp []*snap.AppInfo
+
+func (a bySnapApp) Len() int      { return len(a) }
+func (a bySnapApp) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a bySnapApp) Less(i, j int) bool {
+	iName := a[i].Snap.Name()
+	jName := a[j].Snap.Name()
+	if iName == jName {
+		return a[i].Name < a[j].Name
+	}
+	return iName < jName
+}
+
+func splitAppName(s string) (snap, app string) {
+	if idx := strings.IndexByte(s, '.'); idx > -1 {
+		return s[:idx], s[idx+1:]
+	}
+
+	return s, ""
+}
+
+type wantedAppInfos struct {
+	services bool
+	commands bool
+}
+
+func (w wantedAppInfos) String() string {
+	if w.services {
+		if w.commands {
+			return "app"
+		}
+		return "service"
+	}
+	if !w.commands {
+		logger.Panicf("wanted neither services nor commands")
+	}
+
+	return "command"
+}
+
+// appInfosFor returns a sorted list apps described by names.
+//
+// * If names is empty, returns all apps of the wanted kidns (which
+//   could be an empty list).
+// * An element of names can be a snap name, in which case all apps
+//   from the snap of the wanted kind are included in the result (and
+//   it's an error if the snap has no apps of the wanted kind).
+// * An element of names can instead be snap.app, in which case that app is
```

Comment:
it's a bit strange  to need to use foo.foo to ask for just that app, it's a notation that we never use :/